File: EpicAkS.Blazor.Canvas/JsInterops.cs
using Microsoft.AspNetCore.Components;
using Microsoft.JSInterop;
using System.Text.Json;
using EpicAkS.Blazor.Canvas.Structs;
using System.Security.Cryptography.X509Certificates;

namespace EpicAkS.Blazor.Canvas
{
    internal partial class JsInterops
    {
        private IJSRuntime JSRuntime { get; init; }

        internal JsInterops(IJSRuntime jsRuntime)
        {
            JSRuntime = jsRuntime;
        }

        internal async Task<string> GetMouseXY(int clientX, int clientY)
        {
            if (JSRuntime is null) return String.Empty;
            return await JSRuntime.InvokeAsync<string>("window.epicAkSHelperFunctions.getMouseXY",
                new object[] { clientX, clientY });
        }

        internal async Task<string> GetJSPropertyValuesByElementReference(ElementReference el, string[] props)
        {
            if (JSRuntime is null) return String.Empty;
            return await JSRuntime.InvokeAsync<string>("window.epicAkSHelperFunctions.getElementProps",
                new object[] { el, props });
        }

        internal async Task<string> CreateCanvasForDivElementReferenceWithId(ElementReference elDiv, string canvasId)
        {
            if (JSRuntime is null) return String.Empty;
            return await JSRuntime.InvokeAsync<string>("window.epicAkSHelperFunctions.createCanvasForDivElementReferenceWithId",
                new object[] { elDiv, canvasId });
        }

        internal async Task<string> CreateCanvasForDivElementReferenceWithIdWithContextAttributes(
            ElementReference elDiv, string canvasId, ContextAttributes contextAttributes)
        {
            if (JSRuntime is null) return String.Empty;
            return await JSRuntime.InvokeAsync<string>("window.epicAkSHelperFunctions.createCanvasForDivElementReferenceWithIdWithContextAttributes",
                new object[] { elDiv, canvasId, contextAttributes });
        }

        internal async Task SetCurrentCanvas2DContext(string canvasId)
        {
            if (JSRuntime is null) return;
            await JSRuntime.InvokeVoidAsync("window.epicAkSHelperFunctions.setCurrentCanvas2DContext",
                new object[] { canvasId });
        }

        internal async Task SetCanvas2DContextProperty<T>(string propertyName, T? value)
        {
            if (JSRuntime is null) return;
            await JSRuntime.InvokeVoidAsync("window.epicAkSHelperFunctions.setCanvas2DContextProperty",
                new object?[] { propertyName, value });
        }

        internal async Task SetCanvas2DContextPropertyByExistingVarId<T>(string propertyName, T? value) where T : ReturnVar
        {
            if (JSRuntime is null) return;
            await JSRuntime.InvokeVoidAsync("window.epicAkSHelperFunctions.setCanvas2DContextPropertyByExistingVarId",
                new object?[] { propertyName, (value is not null ? ((ReturnVar)value).VarIdString : null) });
        }

        internal async Task<T?> GetCanvas2DContextProperty<T>(string propertyName)
        {
            if (JSRuntime is null) return default;
            return await JSRuntime.InvokeAsync<T?>("window.epicAkSHelperFunctions.getCanvas2DContextProperty",
                new object[] { propertyName });
        }

        internal async Task<T?> GetCanvas2DContextPropertyWithReturnObjectForProps<T>(string propertyName)
        {
            if (JSRuntime is null) return default;
            string? json = await JSRuntime.InvokeAsync<string?>("window.epicAkSHelperFunctions.getCanvas2DContextPropertyWithReturnObjectForProps",
                new object[] { propertyName, Helpers.FillPropertyNames(typeof(T)) });
            return json is not null ? JsonSerializer.Deserialize<T>(json) : default;
        }

        internal async Task CallCanvas2DContextFunction(string functionName)
        {
            if (JSRuntime is null) return;
            await JSRuntime.InvokeVoidAsync("window.epicAkSHelperFunctions.callCanvas2DContextFunction",
                new object[] { functionName });
        }

        internal async Task CallCanvas2DContextFunctionWithParameters(string functionName, object[] functionParameters)
        {
            if (JSRuntime is null) return;
            await JSRuntime.InvokeVoidAsync("window.epicAkSHelperFunctions.callCanvas2DContextFunctionWithParameters",
                new object[] { functionName, functionParameters });
        }

        internal async Task<T?> CallCanvas2DContextFunctionWithReturn<T>(string functionName)
        {
            if (JSRuntime is null) return default;
            return await JSRuntime.InvokeAsync<T?>("window.epicAkSHelperFunctions.callCanvas2DContextFunctionWithReturn",
                new object[] { functionName });
        }

        internal async Task<T?> CallCanvas2DContextFunctionWithParametersWithReturn<T>(string functionName,
           object[] functionParameters)
        {
            if (JSRuntime is null) return default;
            return await JSRuntime.InvokeAsync<T?>("window.epicAkSHelperFunctions.callCanvas2DContextFunctionWithParametersWithReturn",
                new object[] { functionName, functionParameters });
        }

        internal async Task CallCanvas2DContextFunctionWithExistingVarParameter<T>(string functionName, T tObj) where T : ReturnVar
        {
            if (JSRuntime is null) return;
            if (tObj is not null && !string.IsNullOrWhiteSpace(((ReturnVar)tObj).VarIdString) &&
                !string.IsNullOrWhiteSpace(functionName))
            {
                await JSRuntime.InvokeVoidAsync("window.epicAkSHelperFunctions.callCanvas2DContextFunctionWithExistingVarParameter",
                    new object[] { functionName, ((ReturnVar)tObj).VarIdString });
            }
        }

        internal async Task CallCanvas2DContextFunctionWithExistingVarAndMoreParameters<T>(
           string functionName, T tObj, object[] moreFunctionParameters) where T : ReturnVar
        {
            if (JSRuntime is null) return;
            if (tObj is not null && !string.IsNullOrWhiteSpace(((ReturnVar)tObj).VarIdString) &&
                !string.IsNullOrWhiteSpace(functionName))
            {
                await JSRuntime.InvokeVoidAsync("window.epicAkSHelperFunctions.callCanvas2DContextFunctionWithExistingVarAndMoreParameters",
                    new object[] { functionName, ((ReturnVar)tObj).VarIdString, moreFunctionParameters });
            }
        }

        internal async Task CallCanvas2DContextFunctionWithElementReferenceParameter(string functionName, ElementReference elementReference)
        {
            if (JSRuntime is null) return;
            if (!string.IsNullOrWhiteSpace(functionName))
            {
                await JSRuntime.InvokeVoidAsync("window.epicAkSHelperFunctions.callCanvas2DContextFunctionWithParameters",
                    new object[] { functionName, elementReference });
            }
        }

        internal async Task CallCanvas2DContextFunctionWithElementReferenceAndMoreParameters(string functionName, 
            ElementReference elementReference, object[] moreFunctionParameters)
        {
            if (JSRuntime is null) return;
            if (!string.IsNullOrWhiteSpace(functionName))
            {
                await JSRuntime.InvokeVoidAsync("window.epicAkSHelperFunctions.callCanvas2DContextFunctionWithParameters",
                    (new object[] { functionName, elementReference }).Append(moreFunctionParameters));
            }
        }

        internal async Task CallFunctionWithNewVarToHold<T>(T newTObj, string functionName)
           where T : ReturnVar
        {
            if (JSRuntime is null) return;
            if (newTObj is not null && !string.IsNullOrWhiteSpace(((ReturnVar)newTObj).VarIdString) &&
                !string.IsNullOrWhiteSpace(functionName))
            {
                await JSRuntime.InvokeVoidAsync("window.epicAkSHelperFunctions.callFunctionWithNewVarToHold",
                    new object[] { ((ReturnVar)newTObj).VarIdString, functionName });
            }
        }

        internal async Task CallFunctionWithStringParameterWithNewVarToHold<T>(T newTObj,
           string functionName, string stringParameter)
           where T : ReturnVar
        {
            if (JSRuntime is null) return;
            if (newTObj is not null && !string.IsNullOrWhiteSpace(((ReturnVar)newTObj).VarIdString) &&
                !string.IsNullOrWhiteSpace(functionName))
            {
                await JSRuntime.InvokeVoidAsync("window.epicAkSHelperFunctions.callFunctionWithStringParameterWithNewVarToHold",
                    new object[] { ((ReturnVar)newTObj).VarIdString, functionName, stringParameter });
            }
        }

        internal async Task CallFunctionWithDoubleArrayParameterWithNewVarToHold<T>(T newTObj,
           string functionName, double[] doubleArrayParameter)
           where T : ReturnVar
        {
            if (JSRuntime is null) return;
            if (newTObj is not null && !string.IsNullOrWhiteSpace(((ReturnVar)newTObj).VarIdString) &&
                !string.IsNullOrWhiteSpace(functionName))
            {
                await JSRuntime.InvokeVoidAsync("window.epicAkSHelperFunctions.callFunctionWithArrayParameterWithNewVarToHold",
                    new object[] { ((ReturnVar)newTObj).VarIdString, functionName, doubleArrayParameter });
            }
        }

        internal async Task CallCanvas2DContextFunctionWithVarToHold<T>(T newTObj, string functionName)
           where T : ReturnVar
        {
            if (JSRuntime is null) return;
            if (newTObj is not null && !string.IsNullOrWhiteSpace(((ReturnVar)newTObj).VarIdString) &&
                !string.IsNullOrWhiteSpace(functionName))
            {
                await JSRuntime.InvokeVoidAsync("window.epicAkSHelperFunctions.callCanvas2DContextFunctionWithVarToHold",
                    new object[] { ((ReturnVar)newTObj).VarIdString, functionName });
            }
        }

        internal async Task CallCanvas2DContextFunctionWithParametersWithVarToHold<T>(T newTObj, string functionName,
           object[] functionParameters) where T : ReturnVar
        {
            if (JSRuntime is null) return;
            if (newTObj is not null && !string.IsNullOrWhiteSpace(((ReturnVar)newTObj).VarIdString) &&
                !string.IsNullOrWhiteSpace(functionName) && functionParameters?.Length > 0)
            {
                await JSRuntime.InvokeVoidAsync("window.epicAkSHelperFunctions.callCanvas2DContextFunctionWithParametersWithVarToHold",
                    new object[] { ((ReturnVar)newTObj).VarIdString, functionName, functionParameters });
            }
        }

        internal async Task<T?> CallCanvas2DContextFunctionWithVarToHoldWithReturn<T>(T obj,
           string functionName) where T : ReturnVar
        {
            if (JSRuntime is null) return obj;
            if (obj is not null && !string.IsNullOrWhiteSpace(((ReturnVar)obj).VarIdString) &&
                    ((ReturnVar)obj).PropertyNames?.Length > 0 && !string.IsNullOrWhiteSpace(functionName))
            {
                string json = await JSRuntime.InvokeAsync<string>(
                    "window.epicAkSHelperFunctions.callCanvas2DContextFunctionWithVarToHoldWithReturn",
                    new object[] { ((ReturnVar)obj).VarIdString, functionName, ((ReturnVar)obj).PropertyNames });
                obj = JsonSerializer.Deserialize<T>(json) ?? obj;
            }
            return obj;
        }

        internal async Task<T?> CallCanvas2DContextFunctionWithParametersWithVarToHoldWithReturn<T>(T obj,
           string functionName, object[] functionParameters) where T : ReturnVar
        {
            if (JSRuntime is null) return obj;
            if (obj is not null && !string.IsNullOrWhiteSpace(((ReturnVar)obj).VarIdString) &&
                    ((ReturnVar)obj).PropertyNames?.Length > 0 && !string.IsNullOrWhiteSpace(functionName) &&
                    functionParameters?.Length > 0)
            {
                string json = await JSRuntime.InvokeAsync<string>(
                    "window.epicAkSHelperFunctions.callCanvas2DContextFunctionWithParametersWithVarToHoldWithReturn",
                    new object[] { ((ReturnVar)obj).VarIdString, functionName, functionParameters,
                        ((ReturnVar)obj).PropertyNames });
                obj = JsonSerializer.Deserialize<T>(json) ?? obj;
            }
            return obj;
        }

        internal async Task<T?> GetVarIdPropsFromVarsToHold<T>(T obj) where T : ReturnVar
        {
            if (JSRuntime is null) return obj;
            if (obj is not null && !string.IsNullOrWhiteSpace(((ReturnVar)obj).VarIdString) &&
                    ((ReturnVar)obj).PropertyNames?.Length > 0)
            {
                string json = await JSRuntime.InvokeAsync<string>("window.epicAkSHelperFunctions.getVarIdPropsFromVarsToHold",
                    new object[] { ((ReturnVar)obj).VarIdString, ((ReturnVar)obj).PropertyNames });
                T tObj = JsonSerializer.Deserialize<T>(json) ?? obj;
                if (tObj is not null)
                {
                    ((ReturnVar)tObj).VarId = ((ReturnVar)obj).VarId;
                    ((ReturnVar)tObj).VarIdString = ((ReturnVar)obj).VarIdString;
                    return tObj;
                }
            }
            return obj;
        }

        internal async Task CallFunctionOnExistingVarToHold<T>(T obj, string functionName) where T : ReturnVar
        {
            if (JSRuntime is null) return;
            if (obj is not null && !string.IsNullOrWhiteSpace(((ReturnVar)obj).VarIdString) &&
                !string.IsNullOrWhiteSpace(functionName))
            {
                await JSRuntime.InvokeVoidAsync("window.epicAkSHelperFunctions.callFunctionOnExistingVarToHold",
                    new object[] { ((ReturnVar)obj).VarIdString, functionName });
            }
        }

        internal async Task CallFunctionWithParametersOnExistingVarToHold<T>(T obj,
           string functionName, object[] functionParameters) where T : ReturnVar
        {
            if (JSRuntime is null) return;
            if (obj is not null && !string.IsNullOrWhiteSpace(((ReturnVar)obj).VarIdString) &&
                !string.IsNullOrWhiteSpace(functionName) && functionParameters?.Length > 0)
            {
                await JSRuntime.InvokeVoidAsync("window.epicAkSHelperFunctions.callFunctionWithParametersOnExistingVarToHold",
                    new object[] { ((ReturnVar)obj).VarIdString, functionName, functionParameters });
            }
        }

        internal async Task CallFunctionWithExistingVarParameterOnExistingVarToHold<T1, T2>(T1 t1Obj, string functionName, T2 t2Obj)
           where T1 : ReturnVar where T2 : ReturnVar
        {
            if (JSRuntime is null) return;
            if (t1Obj is not null && !string.IsNullOrWhiteSpace(((ReturnVar)t1Obj).VarIdString) &&
                !string.IsNullOrWhiteSpace(functionName) && t2Obj is not null &&
                !string.IsNullOrWhiteSpace(((ReturnVar)t2Obj).VarIdString))
            {
                await JSRuntime.InvokeVoidAsync("window.epicAkSHelperFunctions.callFunctionWithExistingVarParameterOnExistingVarToHold",
                    new object[] { ((ReturnVar)t1Obj).VarIdString, functionName, ((ReturnVar)t2Obj).VarIdString });
            }
        }
    }
}

File: EpicAkS.Blazor.Canvas/wwwroot/JsInterops.js
window.epicAkSHelperFunctions = {
    ctxByCanvasId: [],
    currentCtx: null,
    varsToHold: [],
    //resizeCanvas: function (canvasId) {
    //    var canvas = document.getElementById(canvasId);
    //    canvas.width = window.innerWidth;
    //    canvas.height = window.innerHeight;

    //    window.onresize = function () {
    //        canvas.width = window.innerWidth;
    //        canvas.height = window.innerHeight;
    //    }
    //},
    getMouseXY: function (clientX, clientY) {
        if(!this.currentCtx) return;
        var rect = this.currentCtx.canvas.getBoundingClientRect();
        return JSON.stringify({
            MouseX: clientX - rect.left + window.pageXOffset,
            MouseY: clientY - rect.top + window.pageYOffset
        });
    },
    getElementProps: function (elem, props) {
        let obj = {};
        for (let i = 0; i < props.length; i++) {
            obj[props[i]] = elem[props[i]];
        }
        return JSON.stringify(obj);
    },
    createCanvasForDivElementReferenceWithId: function (elDiv, canvasId) {
        if (!elDiv || !elDiv.tagName || elDiv.tagName != 'DIV') {
            return;
        }

        let canvas = document.createElement('CANVAS');
        let width = document.documentElement.clientWidth; //window.innerWidth; 
        let height = document.documentElement.clientHeight; //window.innerHeight;
        elDiv.setAttribute('width', width);
        elDiv.setAttribute('height', height);
        canvas.setAttribute('id', canvasId);
        canvas.setAttribute('width', width);
        canvas.setAttribute('height', height);
        elDiv.appendChild(canvas);
        let obj = {};
        obj[canvasId] = canvas.getContext('2d');
        this.ctxByCanvasId.push(obj);
        return '{ "canvasWidth" : ' + width + ', "canvasHeight" : ' + height + ' }';
    },
    createCanvasForDivElementReferenceWithIdWithContextAttributes: function (elDiv, canvasId, contextAttributes) {
        if (!elDiv || !elDiv.tagName || elDiv.tagName != 'DIV') {
            return;
        }

        let canvas = document.createElement('CANVAS');
        let width = window.innerWidth; //document.documentElement.clientWidth;
        let height = window.innerHeight; //document.documentElement.clientHeight;
        elDiv.setAttribute('width', width);
        elDiv.setAttribute('height', height);
        canvas.setAttribute('id', canvasId);
        canvas.setAttribute('width', width);
        canvas.setAttribute('height', height);
        elDiv.appendChild(canvas);
        let obj = {};
        obj[canvasId] = canvas.getContext('2d', contextAttributes);
        this.ctxByCanvasId.push(obj);
        return '{ "canvasWidth" : ' + width + ', "canvasHeight" : ' + height + ' }';
    },
    setCurrentCanvas2DContext: function (canvasId) {
        let ctx;
        for (let i = 0; i < this.ctxByCanvasId.length; i++) {
            ctx = this.ctxByCanvasId[i][canvasId];
            if (ctx) break;
        }
        if (!ctx) return;
        this.currentCtx = ctx;
    },
    setCanvas2DContextProperty: function (propertyName, value) {
        if (!this.currentCtx) return;
        this.currentCtx[propertyName] = value;
    },
    setCanvas2DContextPropertyByExistingVarId: function (propertyName, varId) {
        let varById;
        for (let i = 0; i < this.varsToHold.length; i++) {
            varById = this.varsToHold[i][varId];
            if (varById) break;
        }
        if (!varById) return;
        if (!this.currentCtx || !(propertyName in this.currentCtx)) return
        this.currentCtx[propertyName] = varById;
    },
    getCanvas2DContextProperty: function (propertyName) {
        if (!this.currentCtx || !(propertyName in this.currentCtx)) return null;
        return this.currentCtx[propertyName];
    },
    getCanvas2DContextPropertyWithReturnObjectForProps: function (propertyName, props) {
        if (!this.currentCtx || !(propertyName in this.currentCtx)) return null;
        let obj = this.currentCtx[propertyName];
        let retObj = {};
        for (let i = 0; i < props.length; i++) {
            retObj[props[i]] = obj[props[i]];
        }
        return JSON.stringify(retObj);
    },
    getCanvas2DContextPropertyByExistingVarId: function (propertyName, varId) {
        let varById;
        for (let i = 0; i < this.varsToHold.length; i++) {
            varById = this.varsToHold[i][varId];
            if (varById) break;
        }
        if (!varById) return;
        return varById[propertyName];
    },
    callCanvas2DContextFunction: function (functionName) {
        if (!this.currentCtx || !(functionName in this.currentCtx)) return;
        this.currentCtx[functionName].apply(this.currentCtx);
    },
    callCanvas2DContextFunctionWithParameters: function (functionName, functionParams) {
        if (!this.currentCtx || !(functionName in this.currentCtx)) return;
        this.currentCtx[functionName].apply(this.currentCtx, functionParams);
    },
    callCanvas2DContextFunctionWithReturn: function (functionName) {
        if (!this.currentCtx || !(functionName in this.currentCtx)) return;
        return this.currentCtx[functionName].apply(this.currentCtx);
    },
    callCanvas2DContextFunctionWithParametersWithReturn: function (functionName, functionParams) {
        if (!this.currentCtx || !(functionName in this.currentCtx)) return;
        return this.currentCtx[functionName].apply(this.currentCtx, functionParams);
    },
    callCanvas2DContextFunctionWithExistingVarParameter: function (functionName, functionParamVarId) {
        if (!this.currentCtx || !(functionName in this.currentCtx)) return;
        let functionParamVarById;
        for (let i = 0; i < this.varsToHold.length; i++) {
            functionParamVarById = this.varsToHold[i][functionParamVarId];
            if (functionParamVarById)
                break;
        }
        if (!functionParamVarById) return;
        this.currentCtx[functionName].apply(this.currentCtx, [functionParamVarById]);
    },
    callCanvas2DContextFunctionWithExistingVarAndMoreParameters: function (functionName, functionParamVarId, moreFunctionParameters) {
        if (!this.currentCtx || !(functionName in this.currentCtx)) return;
        let functionParamVarById;
        for (let i = 0; i < this.varsToHold.length; i++) {
            functionParamVarById = this.varsToHold[i][functionParamVarId];
            if (functionParamVarById)
                break;
        }
        if (!functionParamVarById) return;
        this.currentCtx[functionName].apply(this.currentCtx, [functionParamVarById].concat(moreFunctionParameters));
    },
    callCanvas2DContextFunctionWithVarToHold: function (varId, functionName) {
        if (!this.currentCtx || !(functionName in this.currentCtx) || !varId) return;
        let obj = {};
        obj[varId] = this.currentCtx[functionName].apply(this.currentCtx);
        this.varsToHold.push(obj);
    },
    callCanvas2DContextFunctionWithParametersWithVarToHold: function (varId, functionName, functionParams) {
        if (!this.currentCtx || !(functionName in this.currentCtx) || !varId) return;
        let obj = {};
        obj[varId] = this.currentCtx[functionName].apply(this.currentCtx, functionParams);
        this.varsToHold.push(obj);
    },
    callCanvas2DContextFunctionWithVarToHoldWithReturn: function (varId, functionName, props) {
        if (!this.currentCtx || !(functionName in this.currentCtx) || !varId) return;
        let obj = {};
        obj[varId] = this.currentCtx[functionName].apply(this.currentCtx);
        this.varsToHold.push(obj);
        let retObj = {};
        retObj['varId'] = retObj['varIdString'] = varId;
        for (let i = 0; i < props.length; i++) {
            if (props[i] != 'varId' && props[i] != 'varIdString')
                retObj[props[i]] = obj[varId][props[i]];
        }
        return JSON.stringify(retObj);
    },
    callCanvas2DContextFunctionWithParametersWithVarToHoldWithReturn: function (varId, functionName, functionParams, props) {
        if (!this.currentCtx || !(functionName in this.currentCtx) || !varId) return;
        let obj = {};
        obj[varId] = this.currentCtx[functionName].apply(this.currentCtx, functionParams);
        this.varsToHold.push(obj);
        let retObj = {};
        retObj['varId'] = retObj['varIdString'] = varId;
        for (let i = 0; i < props.length; i++) {
            if(props[i] != 'varId' && props[i] != 'varIdString')
                retObj[props[i]] = obj[varId][props[i]]; //varById[props[i]];
        }
        return JSON.stringify(retObj);
    },
    getVarIdPropsFromVarsToHold: function (varId, props) {
        let varById;
        for (let i = 0; i < this.varsToHold.length; i++) {
            varById = this.varsToHold[i][varId];
            if (varById) break;
        }
        if (!varById) return;
        let obj = {};
        for (let i = 0; i < props.length; i++) {
            obj[props[i]] = varById[props[i]];
        }
        return JSON.stringify(obj);
    },
    callFunctionOnExistingVarToHold: function (varId, functionName) {
        let varById;
        for (let i = 0; i < this.varsToHold.length; i++) {
            varById = this.varsToHold[i][varId];
            if (varById) break;
        }
        if (!varById) return;
        varById[functionName].apply(varById);
    },
    callFunctionWithParametersOnExistingVarToHold: function (varId, functionName, functionParams) {
        let varById;
        for (let i = 0; i < this.varsToHold.length; i++) {
            varById = this.varsToHold[i][varId];
            if (varById) break;
        }
        if (!varById) return;
        varById[functionName].apply(varById, functionParams);
    },
    callFunctionWithExistingVarParameterOnExistingVarToHold: function (varId, functionName, functionParamVarId) {
        let varById, functionParamVarById;
        for (let i = 0; i < this.varsToHold.length; i++) {
            if (!varById)
                varById = this.varsToHold[i][varId];
            if (!functionParamVarById)
                functionParamVarById = this.varsToHold[i][functionParamVarId];
        }
        if (!varById && !functionParamVarById) return;
        varById[functionName].apply(varById, [functionParamVarById]);
    },
    callFunctionWithNewVarToHold: function (varId, functionName) {
        let obj = {};
        obj[varId] = new window[functionName]();
        this.varsToHold.push(obj);
    },
    callFunctionWithStringParameterWithNewVarToHold: function (varId, functionName, stringParameter) {
        let obj = {};
        obj[varId] = new window[functionName](stringParameter);
        this.varsToHold.push(obj);
    },
    callFunctionWithArrayParameterWithNewVarToHold: function (varId, functionName, arrayParameter) {
        let obj = {};
        obj[varId] = new window[functionName](arrayParameter);
        this.varsToHold.push(obj);
    }
};
